File: backend/src/utils/period.js
// =====================================================================
// Shared period resolver
// =====================================================================
// One source of truth for "today" / "yesterday" / "this week" / etc.
// Used by every analytics-style endpoint that accepts a `period` query
// param, so the same selection means the same window everywhere.
//
// Returns `{ start, end, prevStart, prevEnd, label }` so callers can
// build period-over-period comparisons without doing date math.
//
// Inspired by the dairy.reports.service version which inlined the
// same logic; that file now imports this util instead.

const DAY_MS = 24 * 60 * 60 * 1000;

const startOfDay = (d) => {
  const x = new Date(d);
  x.setHours(0, 0, 0, 0);
  return x;
};
const endOfDay = (d) => {
  const x = new Date(d);
  x.setHours(23, 59, 59, 999);
  return x;
};

// Sunday-based ISO-like week. Returns the start of the week
// containing `date` (Sun 00:00:00).
const startOfWeek = (date) => {
  const d = startOfDay(date);
  d.setDate(d.getDate() - d.getDay());
  return d;
};

const startOfMonth = (d) =>
  new Date(d.getFullYear(), d.getMonth(), 1);

// Supported preset wire-values. Frontend's PeriodPreset enum mirrors
// this set 1:1 — any change here must be propagated.
export const PERIOD_PRESETS = [
  "today",
  "yesterday",
  "week",
  "lastWeek",
  "month",
  "lastMonth",
  "quarter",
  "halfyear",
  "annual",
  "custom",
];

// Public entrypoint. `period` is the preset; for `custom`, callers
// must also supply `startDate` + `endDate`. Defaults to "month" if
// the preset is missing/unknown so a stale or empty query string
// still resolves to a useful window.
export const resolveRange = ({ period, startDate, endDate } = {}) => {
  const now = new Date();
  const today = startOfDay(now);
  let start;
  let end;
  let label;

  switch (period) {
    case "today": {
      start = startOfDay(today);
      end = endOfDay(today);
      label = "Today";
      break;
    }
    case "yesterday": {
      const y = new Date(today.getTime() - DAY_MS);
      start = startOfDay(y);
      end = endOfDay(y);
      label = "Yesterday";
      break;
    }
    case "week": {
      // Last 7 days ending today, inclusive. Matches how the existing
      // dairy 7-day trend cards bucket data so the chart and the KPI
      // line up.
      end = endOfDay(today);
      start = startOfDay(new Date(today.getTime() - 6 * DAY_MS));
      label = "This week";
      break;
    }
    case "lastWeek": {
      // The 7 days immediately preceding the current 7-day window —
      // pairs with "week" so a side-by-side comparison is symmetric.
      end = endOfDay(new Date(today.getTime() - 7 * DAY_MS));
      start = startOfDay(new Date(today.getTime() - 13 * DAY_MS));
      label = "Last week";
      break;
    }
    case "month": {
      start = startOfMonth(today);
      end = endOfDay(today);
      label = "This month";
      break;
    }
    case "lastMonth": {
      start = new Date(today.getFullYear(), today.getMonth() - 1, 1);
      end = endOfDay(new Date(today.getFullYear(), today.getMonth(), 0));
      label = "Last month";
      break;
    }
    case "quarter": {
      const qStart = Math.floor(today.getMonth() / 3) * 3;
      start = new Date(today.getFullYear(), qStart, 1);
      end = endOfDay(today);
      label = "This quarter";
      break;
    }
    case "halfyear": {
      const hStart = today.getMonth() < 6 ? 0 : 6;
      start = new Date(today.getFullYear(), hStart, 1);
      end = endOfDay(today);
      label = "This half-year";
      break;
    }
    case "annual": {
      start = new Date(today.getFullYear(), 0, 1);
      end = endOfDay(today);
      label = "This year";
      break;
    }
    case "custom": {
      if (!startDate || !endDate) {
        throw new Error("custom period requires startDate and endDate");
      }
      start = startOfDay(new Date(startDate));
      end = endOfDay(new Date(endDate));
      if (end < start) throw new Error("endDate must be on or after startDate");
      label = "Custom range";
      break;
    }
    default: {
      // Default to "month" so an empty / unrecognised query param
      // still returns a useful payload rather than 400'ing.
      start = startOfMonth(today);
      end = endOfDay(today);
      label = "This month";
    }
  }

  // Previous-period window = the same-length window ending the
  // instant before `start`. For preset windows whose semantics
  // already imply a "previous" (e.g. lastWeek pairs with week) this
  // is still the right thing — the comparison is always "this period
  // vs the immediately-preceding span of equal length".
  const lenMs = end.getTime() - start.getTime();
  const prevEnd = new Date(start.getTime() - 1);
  const prevStart = new Date(prevEnd.getTime() - lenMs);
  return {
    start,
    end,
    prevStart: startOfDay(prevStart),
    prevEnd: endOfDay(prevEnd),
    label,
  };
};

// Convenience for endpoints that don't care about previous-period.
export const dateOnly = (range) => ({ start: range.start, end: range.end });

// Compute a delta percentage with the same null-guard semantics the
// dairy reports use: returns `null` (not 0) when `prev` was zero so
// the UI can show "—" instead of a misleading "0%".
export const deltaPct = (current, prev) =>
  prev > 0 ? Number((((current - prev) / prev) * 100).toFixed(1)) : null;

File: frontend/lib/core/util/period.dart
// =====================================================================
// PeriodPreset — global time-period selection
// =====================================================================
// One source of truth, mirroring `backend/src/utils/period.js`. Any
// preset added on either side must be added on the other so the wire
// label stays stable.
//
// Every analytics-style endpoint that takes a `period` query param
// uses these wire-values. The frontend's PeriodFilterPanel renders
// from this enum so adding a preset is one line in two places.

import 'package:intl/intl.dart';

enum PeriodPreset {
  today('today', 'Today'),
  yesterday('yesterday', 'Yesterday'),
  thisWeek('week', 'This Week'),
  lastWeek('lastWeek', 'Last Week'),
  thisMonth('month', 'This Month'),
  lastMonth('lastMonth', 'Last Month'),
  quarter('quarter', 'This Quarter'),
  year('annual', 'This Year'),
  custom('custom', 'Custom');

  const PeriodPreset(this.wire, this.label);

  /// The string sent to the backend as `?period=…`.
  final String wire;
  /// The short label rendered on the filter pill.
  final String label;

  static PeriodPreset fromWire(String wire) =>
      PeriodPreset.values.firstWhere(
        (p) => p.wire == wire,
        orElse: () => PeriodPreset.thisMonth,
      );
}

/// Resolved date range for a preset. Mirrors the backend's
/// `resolveRange()` output so the UI can show "Oct 1 → Oct 15"
/// without round-tripping the server.
class ResolvedRange {
  ResolvedRange({
    required this.start,
    required this.end,
    required this.prevStart,
    required this.prevEnd,
    required this.label,
  });

  final DateTime start;
  final DateTime end;
  final DateTime prevStart;
  final DateTime prevEnd;
  final String label;

  String formattedRange() {
    final fmt = DateFormat('d MMM yy');
    if (start.year == end.year &&
        start.month == end.month &&
        start.day == end.day) {
      return fmt.format(start);
    }
    return '${fmt.format(start)} → ${fmt.format(end)}';
  }
}

DateTime _startOfDay(DateTime d) => DateTime(d.year, d.month, d.day);
DateTime _endOfDay(DateTime d) =>
    DateTime(d.year, d.month, d.day, 23, 59, 59, 999);

/// Frontend-side `resolveRange`. Used for UI labels and any client
/// computations that don't need to round-trip the server.
ResolvedRange resolveRange({
  required PeriodPreset period,
  DateTime? customStart,
  DateTime? customEnd,
}) {
  final now = DateTime.now();
  final today = _startOfDay(now);
  DateTime start;
  DateTime end;
  String label;

  switch (period) {
    case PeriodPreset.today:
      start = _startOfDay(today);
      end = _endOfDay(today);
      label = 'Today';
      break;
    case PeriodPreset.yesterday:
      final y = today.subtract(const Duration(days: 1));
      start = _startOfDay(y);
      end = _endOfDay(y);
      label = 'Yesterday';
      break;
    case PeriodPreset.thisWeek:
      end = _endOfDay(today);
      start = _startOfDay(today.subtract(const Duration(days: 6)));
      label = 'This week';
      break;
    case PeriodPreset.lastWeek:
      end = _endOfDay(today.subtract(const Duration(days: 7)));
      start = _startOfDay(today.subtract(const Duration(days: 13)));
      label = 'Last week';
      break;
    case PeriodPreset.thisMonth:
      start = DateTime(today.year, today.month, 1);
      end = _endOfDay(today);
      label = 'This month';
      break;
    case PeriodPreset.lastMonth:
      start = DateTime(today.year, today.month - 1, 1);
      // Day 0 of current month == last day of previous month.
      end = _endOfDay(DateTime(today.year, today.month, 0));
      label = 'Last month';
      break;
    case PeriodPreset.quarter:
      final qStart = (today.month - 1) ~/ 3 * 3;
      start = DateTime(today.year, qStart + 1, 1);
      end = _endOfDay(today);
      label = 'This quarter';
      break;
    case PeriodPreset.year:
      start = DateTime(today.year, 1, 1);
      end = _endOfDay(today);
      label = 'This year';
      break;
    case PeriodPreset.custom:
      if (customStart == null || customEnd == null) {
        // Fall back to "today" so the panel never crashes on a half-
        // configured custom range; the UI guards against this by
        // requiring both dates before applying the selection.
        start = _startOfDay(today);
        end = _endOfDay(today);
        label = 'Custom (today)';
      } else {
        start = _startOfDay(customStart);
        end = _endOfDay(customEnd);
        label = 'Custom range';
      }
      break;
  }

  final lenMs = end.difference(start).inMilliseconds;
  final prevEnd = _endOfDay(start.subtract(const Duration(milliseconds: 1)));
  final prevStart =
      _startOfDay(prevEnd.subtract(Duration(milliseconds: lenMs)));
  return ResolvedRange(
    start: start,
    end: end,
    prevStart: prevStart,
    prevEnd: prevEnd,
    label: label,
  );
}
